### ISSUE TITLE
The clip_grad_norm_ function in FSDP cannot keep the output consistent when deterministic computing is enabled.

### ISSUE THREAD

**Issue by wangjiangben-hw** [pytorch/pytorch#134393]
### 🐛 Describe the bug

Let's look at this line of the function. 
https://github.com/pytorch/pytorch/blob/main/torch/distributed/fsdp/fully_sharded_data_parallel.py#L1133

Because the set itself is unordered, it is not fixed in the [next function](https://github.com/pytorch/pytorch/blob/main/torch/distributed/fsdp/fully_sharded_data_parallel.py#L2111) for loop after add, which will lead to some deterministic calculation problems. This problem occurs in both sharded_params and non-sharded_params. Please help to fix this problem, thanks~

An easy eaxmple for python set:
```
import multiprocessing

def fn():
    tmp_list = ['a', 'b', 'c', 'd', 'a', 'c']
    print(set(tmp_list))

if __name__ == '__main__':
    p_list = []
    for i in range(10):
        p = multiprocessing.Process(target=fn)
        p_list.append(p)
        p.start()

    for p in p_list:
        p.join()
```

Output is not same between processes:
```
{'a', 'c', 'd', 'b'}
{'b', 'a', 'c', 'd'}
{'b', 'a', 'd', 'c'}
{'a', 'c', 'b', 'd'}
{'a', 'd', 'c', 'b'}
{'b', 'a', 'c', 'd'}
{'c', 'a', 'b', 'd'}
{'d', 'a', 'b', 'c'}
{'d', 'a', 'b', 'c'}
{'b', 'c', 'a', 'd'}
```

### Versions

ANY VERSION

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @zhaojuanmao @mrshenli @rohan-varma @chauhang

**Comment by awgu** [comment]
If you want to open a PR to fix it, I can help review it. Otherwise, give us some time to get bandwidth to do this.

**Comment by zhangheng408** [comment]
this issue is fatal for FSDP hybrid shard. which will result different total_norm and gradients between different replica, then parameters between different replica diverge.  @awgu @albanD @tianyu-l 

```
--- a/torch/distributed/fsdp/fully_sharded_data_parallel.py
+++ b/torch/distributed/fsdp/fully_sharded_data_parallel.py
@@ -1122,18 +1122,27 @@ class FullyShardedDataParallel(nn.Module, _FSDPState):
         max_norm = float(max_norm)
         norm_type = float(norm_type)
         sharded_params = set()
+        sharded_params_list = []
         nonsharded_params = set()  # `NO_SHARD` or not FSDP-managed
+        nonsharded_params_list = []
         grads: List[torch.Tensor] = []
         for handle in self._all_handles:
             target_set = (
                 sharded_params if handle.uses_sharded_strategy else nonsharded_params
             )
+            target_list = (
+                sharded_params_list if handle.uses_sharded_strategy else nonsharded_params_list
+            )
             if handle._use_orig_params:
                 for param in handle.flat_param._params:
+                    if param not in target_set:
+                        target_list.append(param)
                     target_set.add(param)
                     if param.grad is not None:
                         grads.append(param.grad)
             else:
+                if handle.flat_param not in target_set:
+                    target_list.append(handle.flat_param)
                 target_set.add(handle.flat_param)
                 if handle.flat_param.grad is not None:
                     grads.append(handle.flat_param.grad)
@@ -1142,14 +1151,16 @@ class FullyShardedDataParallel(nn.Module, _FSDPState):
                 param not in sharded_params and param not in nonsharded_params
             )
             if not_fsdp_managed:
+                if param not in nonsharded_params:
+                    nonsharded_params_list.append(param)
                 nonsharded_params.add(param)
                 if param.grad is not None:
                     grads.append(param.grad)
         # Compute local norms (forced to be in FP32)
-        local_sharded_norm = _get_grad_norm(sharded_params, norm_type).to(
+        local_sharded_norm = _get_grad_norm(sharded_params_list, norm_type).to(
             self.compute_device
         )
-        local_nonsharded_norm = _get_grad_norm(nonsharded_params, norm_type).to(
+        local_nonsharded_norm = _get_grad_norm(nonsharded_params_list, norm_type).to(
             self.compute_device
         )
         # Reconstruct the total gradient norm depending on the norm type
```
